boa_parser/src/temporal/mod.rs

```diff
@@ -0,0 +1,257 @@
+//! Implementation of ISO8601 grammar lexing/parsing
+use crate::error::ParseResult;
+
+mod annotations;
+mod date_time;
+mod duration;
+mod grammar;
+#[cfg(test)]
+mod tests;
+mod time;
+mod time_zone;
+
+use boa_ast::temporal::{DateRecord, DurationParseRecord, IsoParseRecord, TzIdentifier};
+
+// TODO: optimize where possible.
+//
+// NOTE:
+// Rough max length source given iso calendar and no extraneous annotations
+// is ~100 characters (+10-20 for some calendars):
+// +001970-01-01T00:00:00.000000000+00:00:00.000000000[!America/Argentina/ComodRivadavia][!u-ca=iso8601]
+
+/// Parse a [`TemporalDateTimeString`][proposal].
+///
+/// [proposal]: https://tc39.es/proposal-temporal/#prod-TemporalDateTimeString
+#[derive(Debug, Clone, Copy)]
+pub struct TemporalDateTimeString;
+
+impl TemporalDateTimeString {
+    /// Parses a targeted string as a `DateTime`.
+    ///
+    /// # Errors
+    ///
+    /// The parse will error if the provided target is not valid
+    /// ISO8601 grammar.
+    pub fn parse(zoned: bool, cursor: &mut IsoCursor) -> ParseResult<IsoParseRecord> {
+        date_time::parse_annotated_date_time(zoned, false, false, cursor)
+    }
+}
+
+/// Parse a [`TemporalTimeZoneString`][proposal].
+///
+/// [proposal]: https://tc39.es/proposal-temporal/#prod-TemporalTimeZoneString
+#[derive(Debug, Clone, Copy)]
+pub struct TemporalTimeZoneString;
+
+// TODO: Return a IsoParseRecord instead of a `TzIdentifier`.
+impl TemporalTimeZoneString {
+    /// Parses a targeted string as a `TimeZone`.
+    ///
+    /// # Errors
+    ///
+    /// The parse will error if the provided target is not valid
+    /// ISO8601 grammar.
+    pub fn parse(cursor: &mut IsoCursor) -> ParseResult<TzIdentifier> {
+        time_zone::parse_tz_identifier(cursor)
+    }
+}
+
+/// Parse a [`TemporalYearMonthString`][proposal]
+///
+/// [proposal]: https://tc39.es/proposal-temporal/#prod-TemporalYearMonthString
+#[derive(Debug, Clone, Copy)]
+pub struct TemporalYearMonthString;
+
+impl TemporalYearMonthString {
+    /// Parses a targeted string as a `YearMonth`
+    ///
+    /// # Errors
+    ///
+    /// The parse will error if the provided target is not valid
+    /// ISO8601 grammar.
+    pub fn parse(cursor: &mut IsoCursor) -> ParseResult<IsoParseRecord> {
+        if date_time::peek_year_month(cursor)? {
+            let ym = date_time::parse_year_month(cursor)?;
```

Comment:
Shouldn't this try to parse a YearMonth inconditionally? According to the [spec](https://tc39.es/proposal-temporal/#sec-temporal-parsetemporalyearmonthstring)

> If parseResult is a List of errors, throw a RangeError exception.

Also saw this in `TemporalMonthDayString`, and I think we don't need to double parse the input. The spec does that just so that `ParseISODateTime` fills the default values, but we can impl `Default` constant and do

```rust
let ym = date_time::parse_year_month(cursor)?;

// ... more checks

Ok(
    IsoParseRecord {
        year: ym.0,
        date: ym.1,
        tz_annotation,
        calendar,
        ..Default::default()
    }
)
```

Reply:
I think the reason I was initially peeking for a `YearMonth` / `MonthDay` is due to the record being either a `YearMonth` or an `AnnotatedDateTime`.

[Relevant Link](https://tc39.es/proposal-temporal/#prod-TemporalMonthDayString)

```
TemporalMonthDayString:
    AnnotatedMonthDay
    AnnotatedDateTime
```

I guess we could try to just straight parse `MonthDay` and `YearMonth`, and then if that errors try to parse `AnnotatedDateTime`, but I think peek is relatively quick(?), mostly just confirming general structure at certain indexes... could be wrong on that though.

Reply:
Ahhh, I see, the parser must accept both a month-day or a full date time. This is fine then